// @salutejs/scenario/dist/lib/createSaluteRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSaluteRequest = void 0;
var i18n_1 = require("./i18n");
var createSaluteRequest = function (request) {
    var inference;
    var variant;
    var variables = {};
    return {
        get character() {
            return request.payload.character.id;
        },
        get appInfo() {
            return request.payload.app_info;
        },
        get message() {
            return request.payload.message;
        },
        get systemIntent() {
            return request.payload.intent;
        },
        get variant() {
            return variant;
        },
        get profile() {
            if (request.messageName === 'TAKE_PROFILE_DATA') {
                return request.payload.profile_data;
            }
            return undefined;
        },
        get inference() {
            return inference;
        },
        get request() {
            return request;
        },
        get state() {
            var _a, _b;
            return (_b = (_a = request.payload.meta) === null || _a === void 0 ? void 0 : _a.current_app) === null || _b === void 0 ? void 0 : _b.state;
        },
        get serverAction() {
            // здесь DeprecatedServerAction поскольку используем в SaluteRequest по-умолчанию DeprecatedServerAction
            return request.payload.server_action;
        },
        get voiceAction() {
            return (!request.payload.server_action &&
                request.payload.intent !== 'close_app' &&
                request.payload.original_intent !== 'run_app');
        },
        get variables() {
            return variables;
        },
        setInference: function (value) {
            inference = value;
        },
        setVariable: function (name, value) {
            variables[name] = value;
        },
        i18n: function (keyset) {
            return i18n_1.i18n(request.payload.character.id)(keyset);
        },
        setVariant: function (v) {
            variant = v;
        },
    };
};
exports.createSaluteRequest = createSaluteRequest;
