// @salutejs/scenario/dist/lib/types/payment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PayDialogStatuses = exports.PaymentReponseStatuses = exports.NdsTypes = exports.TaxSystemTypes = exports.PaymentTypes = void 0;
/* eslint-disable no-shadow */
/** Тип оплаты счета */
var PaymentTypes;
(function (PaymentTypes) {
    /** одностадийная оплата */
    PaymentTypes[PaymentTypes["oneStage"] = 0] = "oneStage";
    /** двухстадийная оплата */
    PaymentTypes[PaymentTypes["twoStage"] = 1] = "twoStage";
})(PaymentTypes = exports.PaymentTypes || (exports.PaymentTypes = {}));
/** Система налогообложения */
var TaxSystemTypes;
(function (TaxSystemTypes) {
    /** общая */
    TaxSystemTypes[TaxSystemTypes["obshii"] = 0] = "obshii";
    /** упрощенная, доход */
    TaxSystemTypes[TaxSystemTypes["uproshennui"] = 1] = "uproshennui";
    /** упрощенная, доход минус расход */
    TaxSystemTypes[TaxSystemTypes["dohodMinusRashod"] = 2] = "dohodMinusRashod";
    /** единый налог на вмененный доход */
    TaxSystemTypes[TaxSystemTypes["endfl"] = 3] = "endfl";
    /** единый сельскохозяйственный налог */
    TaxSystemTypes[TaxSystemTypes["esn"] = 4] = "esn";
    /** патентная система налогообложения */
    TaxSystemTypes[TaxSystemTypes["patent"] = 5] = "patent";
})(TaxSystemTypes = exports.TaxSystemTypes || (exports.TaxSystemTypes = {}));
/** Значение ставки НДС */
var NdsTypes;
(function (NdsTypes) {
    /** без НДС */
    NdsTypes[NdsTypes["without"] = 0] = "without";
    /** НДС по ставке 0% */
    NdsTypes[NdsTypes["zeroPercent"] = 1] = "zeroPercent";
    /** НДС по ставке 10% */
    NdsTypes[NdsTypes["tenPercent"] = 2] = "tenPercent";
    /** НДС по ставке 18% */
    NdsTypes[NdsTypes["eighteenPercent"] = 3] = "eighteenPercent";
    /** НДС по ставке 10/110 */
    NdsTypes[NdsTypes["tenToOneTen"] = 4] = "tenToOneTen";
    /** НДС по ставке 18/118 */
    NdsTypes[NdsTypes["eighteenToOneEighteen"] = 5] = "eighteenToOneEighteen";
    /** НДС по ставке 20% */
    NdsTypes[NdsTypes["twentyPercent"] = 6] = "twentyPercent";
    /** НДС по ставке 20/120 */
    NdsTypes[NdsTypes["twentyToOneTwenty"] = 7] = "twentyToOneTwenty";
})(NdsTypes = exports.NdsTypes || (exports.NdsTypes = {}));
var PaymentReponseStatuses;
(function (PaymentReponseStatuses) {
    /** Запрос обработан успешно */
    PaymentReponseStatuses[PaymentReponseStatuses["success"] = 200] = "success";
    /** Один из параметров в запросе передан в некорректном формате, либо формат запроса некорректный */
    PaymentReponseStatuses[PaymentReponseStatuses["wrongRequest"] = 400] = "wrongRequest";
    /** Использован недействительный или неверный API Key (токен) */
    PaymentReponseStatuses[PaymentReponseStatuses["wrongApiKey"] = 401] = "wrongApiKey";
    /** Внутренняя ошибка работы сервиса. Обратитесь в поддержку для устранения неполадки */
    PaymentReponseStatuses[PaymentReponseStatuses["serviceError"] = 403] = "serviceError";
})(PaymentReponseStatuses = exports.PaymentReponseStatuses || (exports.PaymentReponseStatuses = {}));
var PayDialogStatuses;
(function (PayDialogStatuses) {
    /** успешная оплата */
    PayDialogStatuses[PayDialogStatuses["success"] = 0] = "success";
    /** неожиданная ошибка */
    PayDialogStatuses[PayDialogStatuses["unexpectedError"] = 1] = "unexpectedError";
    /**
     * пользователь закрыл смартап.
     * При получении этого кода необходимо дополнительно отправить запрос на проверку статуса платежа,
     * чтобы отобразить результат пользователю */
    PayDialogStatuses[PayDialogStatuses["smarappClosed"] = 2] = "smarappClosed";
    /** невозможно начать оплату, так как отображается другой сценарий оплаты */
    PayDialogStatuses[PayDialogStatuses["anotherPay"] = 3] = "anotherPay";
    /** время оплаты счета истекло */
    PayDialogStatuses[PayDialogStatuses["timesUp"] = 4] = "timesUp";
    /** оплата отклонена бэкендом */
    PayDialogStatuses[PayDialogStatuses["rejected"] = 5] = "rejected";
    /** состояние оплаты неизвестно */
    PayDialogStatuses[PayDialogStatuses["unknown"] = 6] = "unknown";
    /** оплата для данной поверхности недоступна */
    PayDialogStatuses[PayDialogStatuses["unavailable"] = 7] = "unavailable";
})(PayDialogStatuses = exports.PayDialogStatuses || (exports.PayDialogStatuses = {}));
