// @salutejs/client/esm/client-ed11c12c.js
import { _ as __assign, b as __awaiter, c as __generator, d as createNanoEvents, a as __rest } from './common-ba25e019.js';
import { b as commonjsRequire, r as requireMinimal, c as commonjsGlobal, p as protoExports, M as MessageNames, d as createProtocol } from './sdk-7c3cb15e.js';

var asr = {exports: {}};

/*eslint-disable block-scoped-var, id-length, no-control-regex, no-magic-numbers, no-prototype-builtins, no-redeclare, no-shadow, no-var, sort-vars*/

(function (module) {
	(function(global, factory) { /* global define, require, module */

	    /* AMD */ if (typeof commonjsRequire === 'function' && 'object' === 'object' && module && module.exports)
	        module.exports = factory(requireMinimal());

	})(commonjsGlobal, function($protobuf) {

	    // Common aliases
	    var $Reader = $protobuf.Reader, $util = $protobuf.util;
	    
	    // Exported root namespace
	    var $root = $protobuf.roots["default"] || ($protobuf.roots["default"] = {});
	    
	    $root.Variables = (function() {
	    
	        /**
	         * Properties of a Variables.
	         * @exports IVariables
	         * @interface IVariables
	         * @property {Object.<string,string>|null} [variables] Variables variables
	         */
	    
	        /**
	         * Constructs a new Variables.
	         * @exports Variables
	         * @classdesc Represents a Variables.
	         * @implements IVariables
	         * @constructor
	         * @param {IVariables=} [properties] Properties to set
	         */
	        function Variables(properties) {
	            this.variables = {};
	            if (properties)
	                for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
	                    if (properties[keys[i]] != null)
	                        this[keys[i]] = properties[keys[i]];
	        }
	    
	        /**
	         * Variables variables.
	         * @member {Object.<string,string>} variables
	         * @memberof Variables
	         * @instance
	         */
	        Variables.prototype.variables = $util.emptyObject;
	    
	        /**
	         * Decodes a Variables message from the specified reader or buffer.
	         * @function decode
	         * @memberof Variables
	         * @static
	         * @param {$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
	         * @param {number} [length] Message length if known beforehand
	         * @returns {Variables} Variables
	         * @throws {Error} If the payload is not a reader or valid buffer
	         * @throws {$protobuf.util.ProtocolError} If required fields are missing
	         */
	        Variables.decode = function decode(reader, length) {
	            if (!(reader instanceof $Reader))
	                reader = $Reader.create(reader);
	            var end = length === undefined ? reader.len : reader.pos + length, message = new $root.Variables(), key, value;
	            while (reader.pos < end) {
	                var tag = reader.uint32();
	                switch (tag >>> 3) {
	                case 1: {
	                        if (message.variables === $util.emptyObject)
	                            message.variables = {};
	                        var end2 = reader.uint32() + reader.pos;
	                        key = "";
	                        value = "";
	                        while (reader.pos < end2) {
	                            var tag2 = reader.uint32();
	                            switch (tag2 >>> 3) {
	                            case 1:
	                                key = reader.string();
	                                break;
	                            case 2:
	                                value = reader.string();
	                                break;
	                            default:
	                                reader.skipType(tag2 & 7);
	                                break;
	                            }
	                        }
	                        message.variables[key] = value;
	                        break;
	                    }
	                default:
	                    reader.skipType(tag & 7);
	                    break;
	                }
	            }
	            return message;
	        };
	    
	        return Variables;
	    })();
	    
	    $root.UndecodedSeconds = (function() {
	    
	        /**
	         * Properties of an UndecodedSeconds.
	         * @exports IUndecodedSeconds
	         * @interface IUndecodedSeconds
	         * @property {number|null} [undecodedSeconds] UndecodedSeconds undecodedSeconds
	         */
	    
	        /**
	         * Constructs a new UndecodedSeconds.
	         * @exports UndecodedSeconds
	         * @classdesc Represents an UndecodedSeconds.
	         * @implements IUndecodedSeconds
	         * @constructor
	         * @param {IUndecodedSeconds=} [properties] Properties to set
	         */
	        function UndecodedSeconds(properties) {
	            if (properties)
	                for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
	                    if (properties[keys[i]] != null)
	                        this[keys[i]] = properties[keys[i]];
	        }
	    
	        /**
	         * UndecodedSeconds undecodedSeconds.
	         * @member {number} undecodedSeconds
	         * @memberof UndecodedSeconds
	         * @instance
	         */
	        UndecodedSeconds.prototype.undecodedSeconds = 0;
	    
	        /**
	         * Decodes an UndecodedSeconds message from the specified reader or buffer.
	         * @function decode
	         * @memberof UndecodedSeconds
	         * @static
	         * @param {$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
	         * @param {number} [length] Message length if known beforehand
	         * @returns {UndecodedSeconds} UndecodedSeconds
	         * @throws {Error} If the payload is not a reader or valid buffer
	         * @throws {$protobuf.util.ProtocolError} If required fields are missing
	         */
	        UndecodedSeconds.decode = function decode(reader, length) {
	            if (!(reader instanceof $Reader))
	                reader = $Reader.create(reader);
	            var end = length === undefined ? reader.len : reader.pos + length, message = new $root.UndecodedSeconds();
	            while (reader.pos < end) {
	                var tag = reader.uint32();
	                switch (tag >>> 3) {
	                case 1: {
	                        message.undecodedSeconds = reader.float();
	                        break;
	                    }
	                default:
	                    reader.skipType(tag & 7);
	                    break;
	                }
	            }
	            return message;
	        };
	    
	        return UndecodedSeconds;
	    })();
	    
	    $root.FullyFinalized = (function() {
	    
	        /**
	         * Properties of a FullyFinalized.
	         * @exports IFullyFinalized
	         * @interface IFullyFinalized
	         */
	    
	        /**
	         * Constructs a new FullyFinalized.
	         * @exports FullyFinalized
	         * @classdesc Represents a FullyFinalized.
	         * @implements IFullyFinalized
	         * @constructor
	         * @param {IFullyFinalized=} [properties] Properties to set
	         */
	        function FullyFinalized(properties) {
	            if (properties)
	                for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
	                    if (properties[keys[i]] != null)
	                        this[keys[i]] = properties[keys[i]];
	        }
	    
	        /**
	         * Decodes a FullyFinalized message from the specified reader or buffer.
	         * @function decode
	         * @memberof FullyFinalized
	         * @static
	         * @param {$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
	         * @param {number} [length] Message length if known beforehand
	         * @returns {FullyFinalized} FullyFinalized
	         * @throws {Error} If the payload is not a reader or valid buffer
	         * @throws {$protobuf.util.ProtocolError} If required fields are missing
	         */
	        FullyFinalized.decode = function decode(reader, length) {
	            if (!(reader instanceof $Reader))
	                reader = $Reader.create(reader);
	            var end = length === undefined ? reader.len : reader.pos + length, message = new $root.FullyFinalized();
	            while (reader.pos < end) {
	                var tag = reader.uint32();
	                switch (tag >>> 3) {
	                default:
	                    reader.skipType(tag & 7);
	                    break;
	                }
	            }
	            return message;
	        };
	    
	        return FullyFinalized;
	    })();
	    
	    $root.EmotionResult = (function() {
	    
	        /**
	         * Properties of an EmotionResult.
	         * @exports IEmotionResult
	         * @interface IEmotionResult
	         * @property {string|null} [name] EmotionResult name
	         * @property {number|null} [confidence] EmotionResult confidence
	         */
	    
	        /**
	         * Constructs a new EmotionResult.
	         * @exports EmotionResult
	         * @classdesc Represents an EmotionResult.
	         * @implements IEmotionResult
	         * @constructor
	         * @param {IEmotionResult=} [properties] Properties to set
	         */
	        function EmotionResult(properties) {
	            if (properties)
	                for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
	                    if (properties[keys[i]] != null)
	                        this[keys[i]] = properties[keys[i]];
	        }
	    
	        /**
	         * EmotionResult name.
	         * @member {string} name
	         * @memberof EmotionResult
	         * @instance
	         */
	        EmotionResult.prototype.name = "";
	    
	        /**
	         * EmotionResult confidence.
	         * @member {number} confidence
	         * @memberof EmotionResult
	         * @instance
	         */
	        EmotionResult.prototype.confidence = 0;
	    
	        /**
	         * Decodes an EmotionResult message from the specified reader or buffer.
	         * @function decode
	         * @memberof EmotionResult
	         * @static
	         * @param {$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
	         * @param {number} [length] Message length if known beforehand
	         * @returns {EmotionResult} EmotionResult
	         * @throws {Error} If the payload is not a reader or valid buffer
	         * @throws {$protobuf.util.ProtocolError} If required fields are missing
	         */
	        EmotionResult.decode = function decode(reader, length) {
	            if (!(reader instanceof $Reader))
	                reader = $Reader.create(reader);
	            var end = length === undefined ? reader.len : reader.pos + length, message = new $root.EmotionResult();
	            while (reader.pos < end) {
	                var tag = reader.uint32();
	                switch (tag >>> 3) {
	                case 1: {
	                        message.name = reader.string();
	                        break;
	                    }
	                case 2: {
	                        message.confidence = reader.float();
	                        break;
	                    }
	                default:
	                    reader.skipType(tag & 7);
	                    break;
	                }
	            }
	            return message;
	        };
	    
	        return EmotionResult;
	    })();
	    
	    $root.Hypothesis = (function() {
	    
	        /**
	         * Properties of a Hypothesis.
	         * @exports IHypothesis
	         * @interface IHypothesis
	         * @property {string|null} [words] Hypothesis words
	         * @property {number|null} [acousticCost] Hypothesis acousticCost
	         * @property {number|null} [linguisticCost] Hypothesis linguisticCost
	         * @property {number|null} [finalCost] Hypothesis finalCost
	         * @property {number|null} [phraseStart] Hypothesis phraseStart
	         * @property {number|null} [phraseEnd] Hypothesis phraseEnd
	         * @property {string|null} [normalizedText] Hypothesis normalizedText
	         */
	    
	        /**
	         * Constructs a new Hypothesis.
	         * @exports Hypothesis
	         * @classdesc Represents a Hypothesis.
	         * @implements IHypothesis
	         * @constructor
	         * @param {IHypothesis=} [properties] Properties to set
	         */
	        function Hypothesis(properties) {
	            if (properties)
	                for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
	                    if (properties[keys[i]] != null)
	                        this[keys[i]] = properties[keys[i]];
	        }
	    
	        /**
	         * Hypothesis words.
	         * @member {string} words
	         * @memberof Hypothesis
	         * @instance
	         */
	        Hypothesis.prototype.words = "";
	    
	        /**
	         * Hypothesis acousticCost.
	         * @member {number} acousticCost
	         * @memberof Hypothesis
	         * @instance
	         */
	        Hypothesis.prototype.acousticCost = 0;
	    
	        /**
	         * Hypothesis linguisticCost.
	         * @member {number} linguisticCost
	         * @memberof Hypothesis
	         * @instance
	         */
	        Hypothesis.prototype.linguisticCost = 0;
	    
	        /**
	         * Hypothesis finalCost.
	         * @member {number} finalCost
	         * @memberof Hypothesis
	         * @instance
	         */
	        Hypothesis.prototype.finalCost = 0;
	    
	        /**
	         * Hypothesis phraseStart.
	         * @member {number} phraseStart
	         * @memberof Hypothesis
	         * @instance
	         */
	        Hypothesis.prototype.phraseStart = 0;
	    
	        /**
	         * Hypothesis phraseEnd.
	         * @member {number} phraseEnd
	         * @memberof Hypothesis
	         * @instance
	         */
	        Hypothesis.prototype.phraseEnd = 0;
	    
	        /**
	         * Hypothesis normalizedText.
	         * @member {string} normalizedText
	         * @memberof Hypothesis
	         * @instance
	         */
	        Hypothesis.prototype.normalizedText = "";
	    
	        /**
	         * Decodes a Hypothesis message from the specified reader or buffer.
	         * @function decode
	         * @memberof Hypothesis
	         * @static
	         * @param {$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
	         * @param {number} [length] Message length if known beforehand
	         * @returns {Hypothesis} Hypothesis
	         * @throws {Error} If the payload is not a reader or valid buffer
	         * @throws {$protobuf.util.ProtocolError} If required fields are missing
	         */
	        Hypothesis.decode = function decode(reader, length) {
	            if (!(reader instanceof $Reader))
	                reader = $Reader.create(reader);
	            var end = length === undefined ? reader.len : reader.pos + length, message = new $root.Hypothesis();
	            while (reader.pos < end) {
	                var tag = reader.uint32();
	                switch (tag >>> 3) {
	                case 1: {
	                        message.words = reader.string();
	                        break;
	                    }
	                case 2: {
	                        message.acousticCost = reader.float();
	                        break;
	                    }
	                case 3: {
	                        message.linguisticCost = reader.float();
	                        break;
	                    }
	                case 4: {
	                        message.finalCost = reader.float();
	                        break;
	                    }
	                case 5: {
	                        message.phraseStart = reader.float();
	                        break;
	                    }
	                case 6: {
	                        message.phraseEnd = reader.float();
	                        break;
	                    }
	                case 7: {
	                        message.normalizedText = reader.string();
	                        break;
	                    }
	                default:
	                    reader.skipType(tag & 7);
	                    break;
	                }
	            }
	            return message;
	        };
	    
	        return Hypothesis;
	    })();
	    
	    $root.DecoderResult = (function() {
	    
	        /**
	         * Properties of a DecoderResult.
	         * @exports IDecoderResult
	         * @interface IDecoderResult
	         * @property {Array.<IHypothesis>|null} [hypothesis] DecoderResult hypothesis
	         * @property {number|null} [chunkStart] DecoderResult chunkStart
	         * @property {number|null} [chunkEnd] DecoderResult chunkEnd
	         * @property {number|null} [timeEndpointDetectionMs] DecoderResult timeEndpointDetectionMs
	         * @property {number|null} [timeDecodingMs] DecoderResult timeDecodingMs
	         * @property {IVariables|null} [variables] DecoderResult variables
	         * @property {boolean|null} [isFinal] DecoderResult isFinal
	         * @property {Array.<IEmotionResult>|null} [emotionResult] DecoderResult emotionResult
	         * @property {Array.<DecoderResult.IContextAnswer>|null} [contextAnswer] DecoderResult contextAnswer
	         */
	    
	        /**
	         * Constructs a new DecoderResult.
	         * @exports DecoderResult
	         * @classdesc Represents a DecoderResult.
	         * @implements IDecoderResult
	         * @constructor
	         * @param {IDecoderResult=} [properties] Properties to set
	         */
	        function DecoderResult(properties) {
	            this.hypothesis = [];
	            this.emotionResult = [];
	            this.contextAnswer = [];
	            if (properties)
	                for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
	                    if (properties[keys[i]] != null)
	                        this[keys[i]] = properties[keys[i]];
	        }
	    
	        /**
	         * DecoderResult hypothesis.
	         * @member {Array.<IHypothesis>} hypothesis
	         * @memberof DecoderResult
	         * @instance
	         */
	        DecoderResult.prototype.hypothesis = $util.emptyArray;
	    
	        /**
	         * DecoderResult chunkStart.
	         * @member {number} chunkStart
	         * @memberof DecoderResult
	         * @instance
	         */
	        DecoderResult.prototype.chunkStart = 0;
	    
	        /**
	         * DecoderResult chunkEnd.
	         * @member {number} chunkEnd
	         * @memberof DecoderResult
	         * @instance
	         */
	        DecoderResult.prototype.chunkEnd = 0;
	    
	        /**
	         * DecoderResult timeEndpointDetectionMs.
	         * @member {number} timeEndpointDetectionMs
	         * @memberof DecoderResult
	         * @instance
	         */
	        DecoderResult.prototype.timeEndpointDetectionMs = 0;
	    
	        /**
	         * DecoderResult timeDecodingMs.
	         * @member {number} timeDecodingMs
	         * @memberof DecoderResult
	         * @instance
	         */
	        DecoderResult.prototype.timeDecodingMs = 0;
	    
	        /**
	         * DecoderResult variables.
	         * @member {IVariables|null|undefined} variables
	         * @memberof DecoderResult
	         * @instance
	         */
	        DecoderResult.prototype.variables = null;
	    
	        /**
	         * DecoderResult isFinal.
	         * @member {boolean} isFinal
	         * @memberof DecoderResult
	         * @instance
	         */
	        DecoderResult.prototype.isFinal = false;
	    
	        /**
	         * DecoderResult emotionResult.
	         * @member {Array.<IEmotionResult>} emotionResult
	         * @memberof DecoderResult
	         * @instance
	         */
	        DecoderResult.prototype.emotionResult = $util.emptyArray;
	    
	        /**
	         * DecoderResult contextAnswer.
	         * @member {Array.<DecoderResult.IContextAnswer>} contextAnswer
	         * @memberof DecoderResult
	         * @instance
	         */
	        DecoderResult.prototype.contextAnswer = $util.emptyArray;
	    
	        /**
	         * Decodes a DecoderResult message from the specified reader or buffer.
	         * @function decode
	         * @memberof DecoderResult
	         * @static
	         * @param {$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
	         * @param {number} [length] Message length if known beforehand
	         * @returns {DecoderResult} DecoderResult
	         * @throws {Error} If the payload is not a reader or valid buffer
	         * @throws {$protobuf.util.ProtocolError} If required fields are missing
	         */
	        DecoderResult.decode = function decode(reader, length) {
	            if (!(reader instanceof $Reader))
	                reader = $Reader.create(reader);
	            var end = length === undefined ? reader.len : reader.pos + length, message = new $root.DecoderResult();
	            while (reader.pos < end) {
	                var tag = reader.uint32();
	                switch (tag >>> 3) {
	                case 1: {
	                        if (!(message.hypothesis && message.hypothesis.length))
	                            message.hypothesis = [];
	                        message.hypothesis.push($root.Hypothesis.decode(reader, reader.uint32()));
	                        break;
	                    }
	                case 2: {
	                        message.chunkStart = reader.float();
	                        break;
	                    }
	                case 3: {
	                        message.chunkEnd = reader.float();
	                        break;
	                    }
	                case 4: {
	                        message.timeEndpointDetectionMs = reader.float();
	                        break;
	                    }
	                case 5: {
	                        message.timeDecodingMs = reader.float();
	                        break;
	                    }
	                case 6: {
	                        message.variables = $root.Variables.decode(reader, reader.uint32());
	                        break;
	                    }
	                case 7: {
	                        message.isFinal = reader.bool();
	                        break;
	                    }
	                case 8: {
	                        if (!(message.emotionResult && message.emotionResult.length))
	                            message.emotionResult = [];
	                        message.emotionResult.push($root.EmotionResult.decode(reader, reader.uint32()));
	                        break;
	                    }
	                case 9: {
	                        if (!(message.contextAnswer && message.contextAnswer.length))
	                            message.contextAnswer = [];
	                        message.contextAnswer.push($root.DecoderResult.ContextAnswer.decode(reader, reader.uint32()));
	                        break;
	                    }
	                default:
	                    reader.skipType(tag & 7);
	                    break;
	                }
	            }
	            return message;
	        };
	    
	        DecoderResult.ContextAnswer = (function() {
	    
	            /**
	             * Properties of a ContextAnswer.
	             * @memberof DecoderResult
	             * @interface IContextAnswer
	             * @property {Array.<DecoderResult.ContextAnswer.IContextRef>|null} [contextResult] ContextAnswer contextResult
	             */
	    
	            /**
	             * Constructs a new ContextAnswer.
	             * @memberof DecoderResult
	             * @classdesc Represents a ContextAnswer.
	             * @implements IContextAnswer
	             * @constructor
	             * @param {DecoderResult.IContextAnswer=} [properties] Properties to set
	             */
	            function ContextAnswer(properties) {
	                this.contextResult = [];
	                if (properties)
	                    for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
	                        if (properties[keys[i]] != null)
	                            this[keys[i]] = properties[keys[i]];
	            }
	    
	            /**
	             * ContextAnswer contextResult.
	             * @member {Array.<DecoderResult.ContextAnswer.IContextRef>} contextResult
	             * @memberof DecoderResult.ContextAnswer
	             * @instance
	             */
	            ContextAnswer.prototype.contextResult = $util.emptyArray;
	    
	            /**
	             * Decodes a ContextAnswer message from the specified reader or buffer.
	             * @function decode
	             * @memberof DecoderResult.ContextAnswer
	             * @static
	             * @param {$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
	             * @param {number} [length] Message length if known beforehand
	             * @returns {DecoderResult.ContextAnswer} ContextAnswer
	             * @throws {Error} If the payload is not a reader or valid buffer
	             * @throws {$protobuf.util.ProtocolError} If required fields are missing
	             */
	            ContextAnswer.decode = function decode(reader, length) {
	                if (!(reader instanceof $Reader))
	                    reader = $Reader.create(reader);
	                var end = length === undefined ? reader.len : reader.pos + length, message = new $root.DecoderResult.ContextAnswer();
	                while (reader.pos < end) {
	                    var tag = reader.uint32();
	                    switch (tag >>> 3) {
	                    case 1: {
	                            if (!(message.contextResult && message.contextResult.length))
	                                message.contextResult = [];
	                            message.contextResult.push($root.DecoderResult.ContextAnswer.ContextRef.decode(reader, reader.uint32()));
	                            break;
	                        }
	                    default:
	                        reader.skipType(tag & 7);
	                        break;
	                    }
	                }
	                return message;
	            };
	    
	            ContextAnswer.ContextRef = (function() {
	    
	                /**
	                 * Properties of a ContextRef.
	                 * @memberof DecoderResult.ContextAnswer
	                 * @interface IContextRef
	                 * @property {string|null} [id] ContextRef id
	                 * @property {number|null} [index] ContextRef index
	                 * @property {string|null} [originalValue] ContextRef originalValue
	                 * @property {string|null} [predictedValue] ContextRef predictedValue
	                 * @property {number|null} [score] ContextRef score
	                 */
	    
	                /**
	                 * Constructs a new ContextRef.
	                 * @memberof DecoderResult.ContextAnswer
	                 * @classdesc Represents a ContextRef.
	                 * @implements IContextRef
	                 * @constructor
	                 * @param {DecoderResult.ContextAnswer.IContextRef=} [properties] Properties to set
	                 */
	                function ContextRef(properties) {
	                    if (properties)
	                        for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
	                            if (properties[keys[i]] != null)
	                                this[keys[i]] = properties[keys[i]];
	                }
	    
	                /**
	                 * ContextRef id.
	                 * @member {string} id
	                 * @memberof DecoderResult.ContextAnswer.ContextRef
	                 * @instance
	                 */
	                ContextRef.prototype.id = "";
	    
	                /**
	                 * ContextRef index.
	                 * @member {number} index
	                 * @memberof DecoderResult.ContextAnswer.ContextRef
	                 * @instance
	                 */
	                ContextRef.prototype.index = 0;
	    
	                /**
	                 * ContextRef originalValue.
	                 * @member {string} originalValue
	                 * @memberof DecoderResult.ContextAnswer.ContextRef
	                 * @instance
	                 */
	                ContextRef.prototype.originalValue = "";
	    
	                /**
	                 * ContextRef predictedValue.
	                 * @member {string} predictedValue
	                 * @memberof DecoderResult.ContextAnswer.ContextRef
	                 * @instance
	                 */
	                ContextRef.prototype.predictedValue = "";
	    
	                /**
	                 * ContextRef score.
	                 * @member {number} score
	                 * @memberof DecoderResult.ContextAnswer.ContextRef
	                 * @instance
	                 */
	                ContextRef.prototype.score = 0;
	    
	                /**
	                 * Decodes a ContextRef message from the specified reader or buffer.
	                 * @function decode
	                 * @memberof DecoderResult.ContextAnswer.ContextRef
	                 * @static
	                 * @param {$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
	                 * @param {number} [length] Message length if known beforehand
	                 * @returns {DecoderResult.ContextAnswer.ContextRef} ContextRef
	                 * @throws {Error} If the payload is not a reader or valid buffer
	                 * @throws {$protobuf.util.ProtocolError} If required fields are missing
	                 */
	                ContextRef.decode = function decode(reader, length) {
	                    if (!(reader instanceof $Reader))
	                        reader = $Reader.create(reader);
	                    var end = length === undefined ? reader.len : reader.pos + length, message = new $root.DecoderResult.ContextAnswer.ContextRef();
	                    while (reader.pos < end) {
	                        var tag = reader.uint32();
	                        switch (tag >>> 3) {
	                        case 1: {
	                                message.id = reader.string();
	                                break;
	                            }
	                        case 2: {
	                                message.index = reader.int32();
	                                break;
	                            }
	                        case 3: {
	                                message.originalValue = reader.string();
	                                break;
	                            }
	                        case 4: {
	                                message.predictedValue = reader.string();
	                                break;
	                            }
	                        case 5: {
	                                message.score = reader.float();
	                                break;
	                            }
	                        default:
	                            reader.skipType(tag & 7);
	                            break;
	                        }
	                    }
	                    return message;
	                };
	    
	                return ContextRef;
	            })();
	    
	            return ContextAnswer;
	        })();
	    
	        return DecoderResult;
	    })();
	    
	    $root.ErrorResponse = (function() {
	    
	        /**
	         * Properties of an ErrorResponse.
	         * @exports IErrorResponse
	         * @interface IErrorResponse
	         * @property {string|null} [errorMessage] ErrorResponse errorMessage
	         */
	    
	        /**
	         * Constructs a new ErrorResponse.
	         * @exports ErrorResponse
	         * @classdesc Represents an ErrorResponse.
	         * @implements IErrorResponse
	         * @constructor
	         * @param {IErrorResponse=} [properties] Properties to set
	         */
	        function ErrorResponse(properties) {
	            if (properties)
	                for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
	                    if (properties[keys[i]] != null)
	                        this[keys[i]] = properties[keys[i]];
	        }
	    
	        /**
	         * ErrorResponse errorMessage.
	         * @member {string} errorMessage
	         * @memberof ErrorResponse
	         * @instance
	         */
	        ErrorResponse.prototype.errorMessage = "";
	    
	        /**
	         * Decodes an ErrorResponse message from the specified reader or buffer.
	         * @function decode
	         * @memberof ErrorResponse
	         * @static
	         * @param {$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
	         * @param {number} [length] Message length if known beforehand
	         * @returns {ErrorResponse} ErrorResponse
	         * @throws {Error} If the payload is not a reader or valid buffer
	         * @throws {$protobuf.util.ProtocolError} If required fields are missing
	         */
	        ErrorResponse.decode = function decode(reader, length) {
	            if (!(reader instanceof $Reader))
	                reader = $Reader.create(reader);
	            var end = length === undefined ? reader.len : reader.pos + length, message = new $root.ErrorResponse();
	            while (reader.pos < end) {
	                var tag = reader.uint32();
	                switch (tag >>> 3) {
	                case 1: {
	                        message.errorMessage = reader.string();
	                        break;
	                    }
	                default:
	                    reader.skipType(tag & 7);
	                    break;
	                }
	            }
	            return message;
	        };
	    
	        return ErrorResponse;
	    })();
	    
	    $root.PacketWrapperFromServer = (function() {
	    
	        /**
	         * Properties of a PacketWrapperFromServer.
	         * @exports IPacketWrapperFromServer
	         * @interface IPacketWrapperFromServer
	         * @property {IUndecodedSeconds|null} [undecodedSecondsField] PacketWrapperFromServer undecodedSecondsField
	         * @property {IFullyFinalized|null} [fullyFinalizedField] PacketWrapperFromServer fullyFinalizedField
	         * @property {IDecoderResult|null} [decoderResultField] PacketWrapperFromServer decoderResultField
	         * @property {IErrorResponse|null} [errorResponse] PacketWrapperFromServer errorResponse
	         */
	    
	        /**
	         * Constructs a new PacketWrapperFromServer.
	         * @exports PacketWrapperFromServer
	         * @classdesc Represents a PacketWrapperFromServer.
	         * @implements IPacketWrapperFromServer
	         * @constructor
	         * @param {IPacketWrapperFromServer=} [properties] Properties to set
	         */
	        function PacketWrapperFromServer(properties) {
	            if (properties)
	                for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
	                    if (properties[keys[i]] != null)
	                        this[keys[i]] = properties[keys[i]];
	        }
	    
	        /**
	         * PacketWrapperFromServer undecodedSecondsField.
	         * @member {IUndecodedSeconds|null|undefined} undecodedSecondsField
	         * @memberof PacketWrapperFromServer
	         * @instance
	         */
	        PacketWrapperFromServer.prototype.undecodedSecondsField = null;
	    
	        /**
	         * PacketWrapperFromServer fullyFinalizedField.
	         * @member {IFullyFinalized|null|undefined} fullyFinalizedField
	         * @memberof PacketWrapperFromServer
	         * @instance
	         */
	        PacketWrapperFromServer.prototype.fullyFinalizedField = null;
	    
	        /**
	         * PacketWrapperFromServer decoderResultField.
	         * @member {IDecoderResult|null|undefined} decoderResultField
	         * @memberof PacketWrapperFromServer
	         * @instance
	         */
	        PacketWrapperFromServer.prototype.decoderResultField = null;
	    
	        /**
	         * PacketWrapperFromServer errorResponse.
	         * @member {IErrorResponse|null|undefined} errorResponse
	         * @memberof PacketWrapperFromServer
	         * @instance
	         */
	        PacketWrapperFromServer.prototype.errorResponse = null;
	    
	        // OneOf field names bound to virtual getters and setters
	        var $oneOfFields;
	    
	        /**
	         * PacketWrapperFromServer MessageType.
	         * @member {"undecodedSecondsField"|"fullyFinalizedField"|"decoderResultField"|"errorResponse"|undefined} MessageType
	         * @memberof PacketWrapperFromServer
	         * @instance
	         */
	        Object.defineProperty(PacketWrapperFromServer.prototype, "MessageType", {
	            get: $util.oneOfGetter($oneOfFields = ["undecodedSecondsField", "fullyFinalizedField", "decoderResultField", "errorResponse"]),
	            set: $util.oneOfSetter($oneOfFields)
	        });
	    
	        /**
	         * Decodes a PacketWrapperFromServer message from the specified reader or buffer.
	         * @function decode
	         * @memberof PacketWrapperFromServer
	         * @static
	         * @param {$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
	         * @param {number} [length] Message length if known beforehand
	         * @returns {PacketWrapperFromServer} PacketWrapperFromServer
	         * @throws {Error} If the payload is not a reader or valid buffer
	         * @throws {$protobuf.util.ProtocolError} If required fields are missing
	         */
	        PacketWrapperFromServer.decode = function decode(reader, length) {
	            if (!(reader instanceof $Reader))
	                reader = $Reader.create(reader);
	            var end = length === undefined ? reader.len : reader.pos + length, message = new $root.PacketWrapperFromServer();
	            while (reader.pos < end) {
	                var tag = reader.uint32();
	                switch (tag >>> 3) {
	                case 1: {
	                        message.undecodedSecondsField = $root.UndecodedSeconds.decode(reader, reader.uint32());
	                        break;
	                    }
	                case 2: {
	                        message.fullyFinalizedField = $root.FullyFinalized.decode(reader, reader.uint32());
	                        break;
	                    }
	                case 4: {
	                        message.decoderResultField = $root.DecoderResult.decode(reader, reader.uint32());
	                        break;
	                    }
	                case 8: {
	                        message.errorResponse = $root.ErrorResponse.decode(reader, reader.uint32());
	                        break;
	                    }
	                default:
	                    reader.skipType(tag & 7);
	                    break;
	                }
	            }
	            return message;
	        };
	    
	        return PacketWrapperFromServer;
	    })();

	    return $root;
	}); 
} (asr));

var asrExports = asr.exports;

var mtt = {exports: {}};

/*eslint-disable block-scoped-var, id-length, no-control-regex, no-magic-numbers, no-prototype-builtins, no-redeclare, no-shadow, no-var, sort-vars*/

(function (module) {
	(function(global, factory) { /* global define, require, module */

	    /* AMD */ if (typeof commonjsRequire === 'function' && 'object' === 'object' && module && module.exports)
	        module.exports = factory(requireMinimal());

	})(commonjsGlobal, function($protobuf) {

	    // Common aliases
	    var $Reader = $protobuf.Reader, $util = $protobuf.util;
	    
	    // Exported root namespace
	    var $root = $protobuf.roots["default"] || ($protobuf.roots["default"] = {});
	    
	    $root.Music2TrackProtocol = (function() {
	    
	        /**
	         * Namespace Music2TrackProtocol.
	         * @exports Music2TrackProtocol
	         * @namespace
	         */
	        var Music2TrackProtocol = {};
	    
	        Music2TrackProtocol.DecoderResult = (function() {
	    
	            /**
	             * Properties of a DecoderResult.
	             * @memberof Music2TrackProtocol
	             * @interface IDecoderResult
	             * @property {string|null} [result] DecoderResult result
	             * @property {boolean|null} [isMusicFound] DecoderResult isMusicFound
	             * @property {boolean|null} [isFinal] DecoderResult isFinal
	             */
	    
	            /**
	             * Constructs a new DecoderResult.
	             * @memberof Music2TrackProtocol
	             * @classdesc Represents a DecoderResult.
	             * @implements IDecoderResult
	             * @constructor
	             * @param {Music2TrackProtocol.IDecoderResult=} [properties] Properties to set
	             */
	            function DecoderResult(properties) {
	                if (properties)
	                    for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
	                        if (properties[keys[i]] != null)
	                            this[keys[i]] = properties[keys[i]];
	            }
	    
	            /**
	             * DecoderResult result.
	             * @member {string} result
	             * @memberof Music2TrackProtocol.DecoderResult
	             * @instance
	             */
	            DecoderResult.prototype.result = "";
	    
	            /**
	             * DecoderResult isMusicFound.
	             * @member {boolean} isMusicFound
	             * @memberof Music2TrackProtocol.DecoderResult
	             * @instance
	             */
	            DecoderResult.prototype.isMusicFound = false;
	    
	            /**
	             * DecoderResult isFinal.
	             * @member {boolean} isFinal
	             * @memberof Music2TrackProtocol.DecoderResult
	             * @instance
	             */
	            DecoderResult.prototype.isFinal = false;
	    
	            /**
	             * Decodes a DecoderResult message from the specified reader or buffer.
	             * @function decode
	             * @memberof Music2TrackProtocol.DecoderResult
	             * @static
	             * @param {$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
	             * @param {number} [length] Message length if known beforehand
	             * @returns {Music2TrackProtocol.DecoderResult} DecoderResult
	             * @throws {Error} If the payload is not a reader or valid buffer
	             * @throws {$protobuf.util.ProtocolError} If required fields are missing
	             */
	            DecoderResult.decode = function decode(reader, length) {
	                if (!(reader instanceof $Reader))
	                    reader = $Reader.create(reader);
	                var end = length === undefined ? reader.len : reader.pos + length, message = new $root.Music2TrackProtocol.DecoderResult();
	                while (reader.pos < end) {
	                    var tag = reader.uint32();
	                    switch (tag >>> 3) {
	                    case 1: {
	                            message.result = reader.string();
	                            break;
	                        }
	                    case 2: {
	                            message.isMusicFound = reader.bool();
	                            break;
	                        }
	                    case 3: {
	                            message.isFinal = reader.bool();
	                            break;
	                        }
	                    default:
	                        reader.skipType(tag & 7);
	                        break;
	                    }
	                }
	                return message;
	            };
	    
	            return DecoderResult;
	        })();
	    
	        Music2TrackProtocol.ErrorResponse = (function() {
	    
	            /**
	             * Properties of an ErrorResponse.
	             * @memberof Music2TrackProtocol
	             * @interface IErrorResponse
	             * @property {string|null} [errorMessage] ErrorResponse errorMessage
	             * @property {number|null} [errorCode] ErrorResponse errorCode
	             */
	    
	            /**
	             * Constructs a new ErrorResponse.
	             * @memberof Music2TrackProtocol
	             * @classdesc Represents an ErrorResponse.
	             * @implements IErrorResponse
	             * @constructor
	             * @param {Music2TrackProtocol.IErrorResponse=} [properties] Properties to set
	             */
	            function ErrorResponse(properties) {
	                if (properties)
	                    for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
	                        if (properties[keys[i]] != null)
	                            this[keys[i]] = properties[keys[i]];
	            }
	    
	            /**
	             * ErrorResponse errorMessage.
	             * @member {string} errorMessage
	             * @memberof Music2TrackProtocol.ErrorResponse
	             * @instance
	             */
	            ErrorResponse.prototype.errorMessage = "";
	    
	            /**
	             * ErrorResponse errorCode.
	             * @member {number} errorCode
	             * @memberof Music2TrackProtocol.ErrorResponse
	             * @instance
	             */
	            ErrorResponse.prototype.errorCode = 0;
	    
	            /**
	             * Decodes an ErrorResponse message from the specified reader or buffer.
	             * @function decode
	             * @memberof Music2TrackProtocol.ErrorResponse
	             * @static
	             * @param {$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
	             * @param {number} [length] Message length if known beforehand
	             * @returns {Music2TrackProtocol.ErrorResponse} ErrorResponse
	             * @throws {Error} If the payload is not a reader or valid buffer
	             * @throws {$protobuf.util.ProtocolError} If required fields are missing
	             */
	            ErrorResponse.decode = function decode(reader, length) {
	                if (!(reader instanceof $Reader))
	                    reader = $Reader.create(reader);
	                var end = length === undefined ? reader.len : reader.pos + length, message = new $root.Music2TrackProtocol.ErrorResponse();
	                while (reader.pos < end) {
	                    var tag = reader.uint32();
	                    switch (tag >>> 3) {
	                    case 1: {
	                            message.errorMessage = reader.string();
	                            break;
	                        }
	                    case 2: {
	                            message.errorCode = reader.int32();
	                            break;
	                        }
	                    default:
	                        reader.skipType(tag & 7);
	                        break;
	                    }
	                }
	                return message;
	            };
	    
	            return ErrorResponse;
	        })();
	    
	        Music2TrackProtocol.MttResponse = (function() {
	    
	            /**
	             * Properties of a MttResponse.
	             * @memberof Music2TrackProtocol
	             * @interface IMttResponse
	             * @property {Music2TrackProtocol.IDecoderResult|null} [decoderResultField] MttResponse decoderResultField
	             * @property {Music2TrackProtocol.IErrorResponse|null} [errorResponse] MttResponse errorResponse
	             */
	    
	            /**
	             * Constructs a new MttResponse.
	             * @memberof Music2TrackProtocol
	             * @classdesc Represents a MttResponse.
	             * @implements IMttResponse
	             * @constructor
	             * @param {Music2TrackProtocol.IMttResponse=} [properties] Properties to set
	             */
	            function MttResponse(properties) {
	                if (properties)
	                    for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
	                        if (properties[keys[i]] != null)
	                            this[keys[i]] = properties[keys[i]];
	            }
	    
	            /**
	             * MttResponse decoderResultField.
	             * @member {Music2TrackProtocol.IDecoderResult|null|undefined} decoderResultField
	             * @memberof Music2TrackProtocol.MttResponse
	             * @instance
	             */
	            MttResponse.prototype.decoderResultField = null;
	    
	            /**
	             * MttResponse errorResponse.
	             * @member {Music2TrackProtocol.IErrorResponse|null|undefined} errorResponse
	             * @memberof Music2TrackProtocol.MttResponse
	             * @instance
	             */
	            MttResponse.prototype.errorResponse = null;
	    
	            // OneOf field names bound to virtual getters and setters
	            var $oneOfFields;
	    
	            /**
	             * MttResponse MessageType.
	             * @member {"decoderResultField"|"errorResponse"|undefined} MessageType
	             * @memberof Music2TrackProtocol.MttResponse
	             * @instance
	             */
	            Object.defineProperty(MttResponse.prototype, "MessageType", {
	                get: $util.oneOfGetter($oneOfFields = ["decoderResultField", "errorResponse"]),
	                set: $util.oneOfSetter($oneOfFields)
	            });
	    
	            /**
	             * Decodes a MttResponse message from the specified reader or buffer.
	             * @function decode
	             * @memberof Music2TrackProtocol.MttResponse
	             * @static
	             * @param {$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
	             * @param {number} [length] Message length if known beforehand
	             * @returns {Music2TrackProtocol.MttResponse} MttResponse
	             * @throws {Error} If the payload is not a reader or valid buffer
	             * @throws {$protobuf.util.ProtocolError} If required fields are missing
	             */
	            MttResponse.decode = function decode(reader, length) {
	                if (!(reader instanceof $Reader))
	                    reader = $Reader.create(reader);
	                var end = length === undefined ? reader.len : reader.pos + length, message = new $root.Music2TrackProtocol.MttResponse();
	                while (reader.pos < end) {
	                    var tag = reader.uint32();
	                    switch (tag >>> 3) {
	                    case 1: {
	                            message.decoderResultField = $root.Music2TrackProtocol.DecoderResult.decode(reader, reader.uint32());
	                            break;
	                        }
	                    case 2: {
	                            message.errorResponse = $root.Music2TrackProtocol.ErrorResponse.decode(reader, reader.uint32());
	                            break;
	                        }
	                    default:
	                        reader.skipType(tag & 7);
	                        break;
	                    }
	                }
	                return message;
	            };
	    
	            return MttResponse;
	        })();
	    
	        return Music2TrackProtocol;
	    })();

	    return $root;
	}); 
} (mtt));

var mttExports = mtt.exports;

var createClient$1 = function (protocol, provideMeta, _a) {
    if (provideMeta === void 0) { provideMeta = undefined; }
    var getVoiceMeta = _a.getVoiceMeta;
    var _b = createNanoEvents(), on = _b.on, emit = _b.emit;
    /** ждет ответ бека и возвращает данные из этого ответа */
    var waitForAnswer = function (messageId) {
        return new Promise(function (resolve) {
            var off = on('systemMessage', function (systemMessageData, originalMessage) {
                if (originalMessage.messageId === messageId &&
                    (originalMessage.messageName === MessageNames.ANSWER_TO_USER ||
                        originalMessage.messageName === MessageNames.DO_NOTHING)) {
                    off();
                    resolve(systemMessageData);
                }
            });
        });
    };
    /** отправляет произвольный systemMessage, не подкладывает мету */
    var sendData = function (data, messageName, meta) {
        if (messageName === void 0) { messageName = ''; }
        var messageId = protocol.getMessageId();
        protocol.sendSystemMessage({
            data: data,
            messageName: messageName,
        }, true, messageId, { meta: meta || {} });
        return messageId;
    };
    /** отправляет cancel на сообщение */
    var sendCancel = function (messageId) {
        protocol.sendCancel({}, true, messageId);
    };
    /** отправляет приветствие */
    var sendOpenAssistant = function (_a) {
        var _b = _a === void 0 ? { isFirstSession: false } : _a, isFirstSession = _b.isFirstSession;
        return __awaiter(void 0, void 0, void 0, function () {
            var data, meta, _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        data = isFirstSession ? { is_first_session: true } : {};
                        if (!provideMeta) return [3 /*break*/, 2];
                        return [4 /*yield*/, provideMeta()];
                    case 1:
                        _c = _d.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        _c = undefined;
                        _d.label = 3;
                    case 3:
                        meta = _c;
                        return [2 /*return*/, waitForAnswer(sendData(data, 'OPEN_ASSISTANT', meta))];
                }
            });
        });
    };
    /** вызывает sendSystemMessage, куда подкладывает мету */
    var sendMeta = function (sendSystemMessage, additionalMeta) { return __awaiter(void 0, void 0, void 0, function () {
        var meta, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    if (!provideMeta) return [3 /*break*/, 2];
                    return [4 /*yield*/, provideMeta(additionalMeta)];
                case 1:
                    _a = _b.sent();
                    return [3 /*break*/, 3];
                case 2:
                    _a = {};
                    _b.label = 3;
                case 3:
                    meta = _a;
                    if (typeof meta !== 'undefined') {
                        sendSystemMessage({
                            data: {},
                            messageName: '',
                        }, false, {
                            meta: meta,
                        });
                    }
                    return [2 /*return*/];
            }
        });
    }); };
    /** перестать слать озвучку для messageId */
    var sendMute = function (messageId) {
        protocol.sendMute({}, true, messageId);
    };
    /** отправляет server_action и мету */
    var sendServerAction = function (serverAction, appInfo, messageName, mode) {
        if (messageName === void 0) { messageName = 'SERVER_ACTION'; }
        return __awaiter(void 0, void 0, void 0, function () {
            var messageId;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        messageId = protocol.getMessageId();
                        // мету и server_action отправляем в одном systemMessage
                        return [4 /*yield*/, sendMeta(function (data, _, _a) {
                                var _b = _a === void 0 ? {} : _a, meta = _b.meta;
                                var systemData = __rest(data, []);
                                protocol.sendSystemMessage({
                                    // eslint-disable-next-line camelcase
                                    data: __assign(__assign({}, systemData), { app_info: appInfo, server_action: serverAction, mode: mode }),
                                    messageName: messageName || 'SERVER_ACTION',
                                }, true, messageId, { meta: meta });
                            }, {
                                source: {
                                    sourceType: 'vps',
                                },
                            })];
                    case 1:
                        // мету и server_action отправляем в одном systemMessage
                        _a.sent();
                        return [2 /*return*/, messageId];
                }
            });
        });
    };
    /** отправляет текст и текущую мету */
    var sendText = function (text, isSsml, shouldSendDisableDubbing, additionalMeta) {
        if (isSsml === void 0) { isSsml = false; }
        return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (text.trim() === '') {
                    return [2 /*return*/, undefined];
                }
                return [2 /*return*/, protocol.batch(function (_a) {
                        var sendSystemMessage = _a.sendSystemMessage, clientSendText = _a.sendText, sendSettings = _a.sendSettings, messageId = _a.messageId;
                        return __awaiter(void 0, void 0, void 0, function () {
                            var prevDubbing, sendDisableDubbing, isStillNeedReturnDubbing;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0: return [4 /*yield*/, sendMeta(sendSystemMessage, additionalMeta)];
                                    case 1:
                                        _b.sent();
                                        prevDubbing = protocol.configuration.settings.dubbing;
                                        sendDisableDubbing = prevDubbing !== -1 && shouldSendDisableDubbing;
                                        if (!sendDisableDubbing) return [3 /*break*/, 3];
                                        return [4 /*yield*/, sendSettings({ dubbing: -1 }, false)];
                                    case 2:
                                        _b.sent();
                                        _b.label = 3;
                                    case 3:
                                        isSsml ? clientSendText(text, {}, 'application/ssml') : clientSendText(text, {});
                                        isStillNeedReturnDubbing = prevDubbing === protocol.configuration.settings.dubbing;
                                        if (sendDisableDubbing && isStillNeedReturnDubbing) {
                                            sendSettings({ dubbing: prevDubbing });
                                        }
                                        return [2 /*return*/, messageId];
                                }
                            });
                        });
                    })];
            });
        });
    };
    /** инициализирует исходящий голосовой поток, факт. передает в callback параметры для отправки голоса,
     * отправляет мету */
    var createVoiceStream = function (callback, additionalMeta) {
        return protocol.batch(function (_a) {
            var sendSystemMessage = _a.sendSystemMessage, sendVoice = _a.sendVoice, messageId = _a.messageId;
            return __awaiter(void 0, void 0, void 0, function () {
                var first;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            first = true;
                            return [4 /*yield*/, sendMeta(sendSystemMessage, additionalMeta)];
                        case 1:
                            _b.sent();
                            return [4 /*yield*/, callback({
                                    sendVoice: function () {
                                        var args = [];
                                        for (var _i = 0; _i < arguments.length; _i++) {
                                            args[_i] = arguments[_i];
                                        }
                                        if (first) {
                                            // отправляем мету только в первом чанке с голосом
                                            sendVoice(args[0], args[1], args[2], __assign(__assign({}, args[3]), { meta: getVoiceMeta() }));
                                            first = false;
                                            return;
                                        }
                                        sendVoice.apply(void 0, args);
                                    },
                                    messageId: messageId,
                                })];
                        case 2:
                            _b.sent();
                            return [2 /*return*/];
                    }
                });
            });
        });
    };
    var off = [
        protocol.on('incoming', function (message) {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j;
            if ((_a = message.systemMessage) === null || _a === void 0 ? void 0 : _a.data) {
                emit('systemMessage', JSON.parse(message.systemMessage.data), message);
            }
            if (message.voice) {
                emit('voice', message.voice.data || new Uint8Array(), message);
            }
            if (message.status) {
                emit('status', message.status, message);
            }
            if (message.messageName === 'TAKE_HISTORY' && ((_b = message.bytes) === null || _b === void 0 ? void 0 : _b.data)) {
                var history_1 = protoExports.GetHistoryResponse.decode((_c = message.bytes) === null || _c === void 0 ? void 0 : _c.data).historyMessages;
                var parsedHistory = history_1.map(function (historyMessage) { return (__assign(__assign({}, historyMessage), { content: JSON.parse(historyMessage.content || '') })); });
                emit('history', parsedHistory, message);
            }
            if (message.messageName === MessageNames.STT && (message.text || ((_e = (_d = message.bytes) === null || _d === void 0 ? void 0 : _d.data) === null || _e === void 0 ? void 0 : _e.length))) {
                var response = ((_g = (_f = message.bytes) === null || _f === void 0 ? void 0 : _f.data) === null || _g === void 0 ? void 0 : _g.length)
                    ? asrExports.PacketWrapperFromServer.decode(message.bytes.data)
                    : undefined;
                emit('stt', { text: message.text, response: response }, message);
            }
            if (message.messageName === MessageNames.MTT && ((_j = (_h = message.bytes) === null || _h === void 0 ? void 0 : _h.data) === null || _j === void 0 ? void 0 : _j.length)) {
                emit('musicRecognition', mttExports.Music2TrackProtocol.MttResponse.decode(message.bytes.data), message);
            }
        }),
        protocol.on('ready', function () {
            emit('ready');
        }),
    ];
    return {
        destroy: function () {
            off.map(function (f) { return f(); });
        },
        close: protocol.close,
        init: protocol.init,
        createVoiceStream: createVoiceStream,
        sendData: sendData,
        sendMeta: sendMeta,
        sendOpenAssistant: sendOpenAssistant,
        sendServerAction: sendServerAction,
        sendText: sendText,
        sendCancel: sendCancel,
        sendMute: sendMute,
        on: on,
        waitForAnswer: waitForAnswer,
        get status() {
            return protocol.status;
        },
        get currentMessageId() {
            return protocol.currentMessageId;
        },
    };
};

var RETRY_INTERVAL = 300; // ms
var defaultWSCreator = function (url) { return new WebSocket(url); };
var createTransport = function (_a) {
    var _b = _a.createWS, createWS = _b === void 0 ? defaultWSCreator : _b, checkCertUrl = _a.checkCertUrl;
    var _c = createNanoEvents(), on = _c.on, emit = _c.emit;
    var hasCert = !checkCertUrl;
    var retryTimeoutId = -1;
    var retries = 0;
    var status = 'closed';
    var webSocket;
    var stopped = true;
    var checkCert = function (checkUrl) {
        return new Promise(function (resolve) {
            window
                .fetch(checkUrl)
                .then(function () { return resolve(true); })
                .catch(function () { return resolve(false); });
        });
    };
    var close = function () {
        stopped = true;
        if (status === 'closing' || status === 'closed') {
            return;
        }
        status = 'closing';
        webSocket === null || webSocket === void 0 ? void 0 : webSocket.close();
    };
    var connect = function (url) { return __awaiter(void 0, void 0, void 0, function () {
        var okay;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    status = 'connecting';
                    emit('connecting');
                    if (!(!hasCert && window.navigator.onLine)) return [3 /*break*/, 2];
                    return [4 /*yield*/, checkCert(checkCertUrl)];
                case 1:
                    okay = _a.sent();
                    if (!okay) {
                        status = 'closed';
                        emit('close');
                        emit('error', new Error('Cert authority invalid'));
                        return [2 /*return*/];
                    }
                    hasCert = true;
                    _a.label = 2;
                case 2:
                    webSocket = createWS(url);
                    webSocket.binaryType = 'arraybuffer';
                    webSocket.addEventListener('open', function () {
                        if (webSocket.readyState !== 1) {
                            return;
                        }
                        window.clearTimeout(retryTimeoutId);
                        retries = 0;
                        status = 'open';
                        emit('open');
                    });
                    webSocket.addEventListener('close', function () {
                        status = 'closed';
                        emit('close');
                    });
                    webSocket.addEventListener('error', function (e) {
                        if (status !== 'connecting') {
                            throw e;
                        }
                        // пробуем переподключаться, если возникла ошибка при коннекте
                        if (!webSocket || (webSocket.readyState === 3 && !stopped)) {
                            window.clearTimeout(retryTimeoutId);
                            if (retries < 2) {
                                retryTimeoutId = window.setTimeout(function () {
                                    // eslint-disable-next-line @typescript-eslint/no-use-before-define
                                    open(url);
                                    retries++;
                                }, RETRY_INTERVAL * retries);
                            }
                            else {
                                retries = 0;
                                emit('error', e);
                            }
                        }
                    });
                    webSocket.addEventListener('message', function (_a) {
                        var data = _a.data;
                        emit('message', data);
                    });
                    return [2 /*return*/];
            }
        });
    }); };
    var open = function (url) {
        if (status === 'connecting' || status === 'open') {
            return;
        }
        stopped = false;
        connect(url);
    };
    var reconnect = function (url) {
        if (status === 'closed') {
            open(url);
            return;
        }
        window.setTimeout(function () { return reconnect(url); });
        close();
    };
    var send = function (data) {
        if (!window.navigator.onLine) {
            close();
            emit('error');
            throw new Error('The client seems to be offline');
        }
        webSocket.send(data);
    };
    return {
        close: close,
        get isOnline() {
            return window.navigator.onLine;
        },
        on: on,
        open: open,
        reconnect: reconnect,
        send: send,
    };
};

var convertToMetaPermissions = function (permission) {
    return Object.keys(permission).map(function (key) { return ({
        type: key,
        status: permission[key],
    }); });
};
var getLocation = function () { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        return [2 /*return*/, new Promise(function (resolve, reject) {
                navigator.geolocation.getCurrentPosition(function (_a) {
                    var coords = _a.coords, timestamp = _a.timestamp;
                    resolve({
                        lat: coords.latitude,
                        lon: coords.longitude,
                        accuracy: coords.accuracy,
                        timestamp: timestamp,
                    });
                }, reject, { timeout: 5000 });
            })];
    });
}); };
var getTime = function () { return ({
    // Здесь нужен полифилл, т.к. `Intl.DateTimeFormat().resolvedOptions().timeZone` - возвращает пустую строку
    timezone_id: Intl.DateTimeFormat().resolvedOptions().timeZone,
    timezone_offset_sec: -new Date().getTimezoneOffset() * 60,
    timestamp: Date.now(),
}); };
var getAnswerForRequestPermissions = function (requestMessageId, appInfo, items) { return __awaiter(void 0, void 0, void 0, function () {
    var permissions, response;
    return __generator(this, function (_a) {
        permissions = {
            record_audio: 'denied_once',
            geo: 'denied_once',
            read_contacts: 'denied_permanently',
            push: 'denied_once',
        };
        response = {
            app_info: appInfo,
            meta: {
                time: getTime(),
                permissions: [],
            },
            server_action: {
                action_id: 'command_response',
                request_message_id: requestMessageId,
                command_response: {
                    request_permissions: {
                        permissions: [],
                    },
                },
            },
        };
        return [2 /*return*/, Promise.all(items.map(function (permission) { return __awaiter(void 0, void 0, void 0, function () {
                var _a, location_1;
                var _c, _d, _e;
                return __generator(this, function (_f) {
                    switch (_f.label) {
                        case 0:
                            _a = permission;
                            switch (_a) {
                                case 'geo': return [3 /*break*/, 1];
                            }
                            return [3 /*break*/, 5];
                        case 1:
                            _f.trys.push([1, 3, , 4]);
                            return [4 /*yield*/, getLocation()];
                        case 2:
                            location_1 = _f.sent();
                            permissions.geo = 'granted';
                            response.meta.location = location_1;
                            (_c = response.server_action.command_response.request_permissions) === null || _c === void 0 ? void 0 : _c.permissions.push({
                                type: 'geo',
                                status: 'granted',
                            });
                            return [3 /*break*/, 4];
                        case 3:
                            _f.sent();
                            permissions.geo = 'denied_permanently';
                            (_d = response.server_action.command_response.request_permissions) === null || _d === void 0 ? void 0 : _d.permissions.push({
                                type: 'geo',
                                status: 'denied_permanently',
                            });
                            return [3 /*break*/, 4];
                        case 4: return [3 /*break*/, 6];
                        case 5:
                            // остальные доступы не поддерживаем
                            (_e = response.server_action.command_response.request_permissions) === null || _e === void 0 ? void 0 : _e.permissions.push({
                                type: permission,
                                status: 'denied_permanently',
                            });
                            return [3 /*break*/, 6];
                        case 6: return [2 /*return*/];
                    }
                });
            }); })).then(function () {
                response.meta.permissions = convertToMetaPermissions(permissions);
                return response;
            })];
    });
}); };

var STATE_UPDATE_TIMEOUT = 200;
var DEFAULT_PROJECT_ID = 'd929986a-611a-2ba0-6174-1928c99600a5';
var DEFAULT_APPLICATION_ID = '7c4e23bf-cd93-b57e-874b-d9fc1b35f93d';
var DEFAULT_APP_VERSION_ID = '26d0bb2e-45d6-a276-f70e-6c016d1f9cff';
var DEFAULT_APP = {
    projectId: DEFAULT_PROJECT_ID,
    applicationId: DEFAULT_APPLICATION_ID,
    appversionId: DEFAULT_APP_VERSION_ID,
    frontendStateId: [DEFAULT_PROJECT_ID, DEFAULT_APPLICATION_ID, DEFAULT_APP_VERSION_ID].join('_'),
    frontendType: 'DIALOG',
    systemName: 'assistant',
    frontendEndpoint: 'None',
};
var BASIC_SMART_APP_COMMANDS_TYPES = ['smart_app_data', 'smart_app_error', 'start_smart_search', 'navigation'];

function convertFieldValuesToString(object) {
    return Object.keys(object).reduce(function (acc, key) {
        if (object[key]) {
            acc[key] =
                typeof object[key] === 'string' && object[key].startsWith('{')
                    ? object[key]
                    : JSON.stringify(object[key]);
        }
        return acc;
    }, {});
}
var isDefaultApp = function (appInfo) { return appInfo.frontendStateId === DEFAULT_APP.frontendStateId; };
var promiseTimeout = function (promise, timeout) {
    var timeoutId;
    return Promise.race([
        promise.then(function (v) {
            if (timeoutId) {
                window.clearTimeout(timeoutId);
            }
            return v;
        }),
        new Promise(function (_, reject) {
            timeoutId = window.setTimeout(function () {
                reject(new Error("Timed out in " + timeout + " ms."));
            }, timeout);
        }),
    ]);
};

var createClient = function (configuration) {
    var metaProvider = function () { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, Promise.resolve(convertFieldValuesToString({
                    theme: 'dark',
                    time: getTime(),
                    // eslint-disable-next-line camelcase
                    current_app: {
                        // eslint-disable-next-line camelcase
                        app_info: DEFAULT_APP,
                    },
                }))];
        });
    }); };
    var transport = createTransport({});
    var protocol = createProtocol(transport, __assign(__assign({}, configuration), { getInitialMeta: function () {
            return convertFieldValuesToString({
                default_character: 'sber',
            });
        } }));
    return createClient$1(protocol, metaProvider, {
        getVoiceMeta: function () { return ({}); },
    });
};

export { BASIC_SMART_APP_COMMANDS_TYPES as B, DEFAULT_APP as D, STATE_UPDATE_TIMEOUT as S, asrExports as a, createTransport as b, createClient as c, convertFieldValuesToString as d, createClient$1 as e, getAnswerForRequestPermissions as f, getTime as g, isDefaultApp as i, promiseTimeout as p };
