// @salutejs/client/esm/index.js
export { createAssistant } from './createAssistant.js';
export { createAssistantHostMock, createAssistantHostMockWithRecord } from './mock.js';
export { e as createAnswerFromMockByMessageGetter, f as createMockRecorder, d as createMockWSCreator, b as createOnlineRecordPlayer, a as createRecordPlayer } from './record-0905f37a.js';
export { createAssistantDev, createSmartappDebugger } from './createAssistantDev.js';
export { c as createAssistantClient, c as createAssistantSDK, b as createAudioRecorder, a as createNavigatorAudioProvider } from './assistant-c0a66ca1.js';
import { p as protoExports } from './sdk-7c3cb15e.js';
export { M as MessageNames, V as VpsVersion } from './sdk-7c3cb15e.js';
import { a as asrExports } from './client-ed11c12c.js';
export { c as createClient } from './client-ed11c12c.js';
export { c as createVoicePlayer } from './voicePlayer-705b49ef.js';
import './common-ba25e019.js';



var GetHistoryResponse = protoExports.GetHistoryResponse;
var InitialSettings = protoExports.InitialSettings;
var PacketWrapperFromServer = asrExports.PacketWrapperFromServer;
export { GetHistoryResponse, InitialSettings, PacketWrapperFromServer };
